platform/internal/mbed_os_timer.cpp

```diff
@@ -0,0 +1,243 @@
+            CriticalSectionLock lock;
+
+            // Recheck wake conditions before starting sleep, avoiding race
+            if (op.wake_condition()) {
+                break;
+            }
+
+            // It's possible that an intermediate wake interrupt occurred
+            // between "set_wake_time" and the critical lock - only sleep
+            // if we see that the timer is armed or we don't need it. Otherwise,
+            // we go round to set the timer again.
+            if (op.sleep_prepared()) {
+                // Enter HAL sleep (normal or deep)
+                sleep();
+            }
+        }
+
+        // Ensure interrupts get a chance to fire, which allows new result from
+        // wake_predicate() and wake_time_passed()
+        __ISB();
+    }
+}
+
+/* We require that this is called from thread context, outside a critical section,
+ * and the kernel already suspended if an RTOS, meaning we don't have to worry
+ * about any potential threading issues.
+ *
+ * The wake predicate will be called from both outside and inside a critical
+ * section, so appropriate atomic care must be taken.
+ */
+uint64_t do_timed_sleep_absolute(uint64_t wake_time, bool (*wake_predicate)(void *), void *wake_predicate_handle)
+{
+    {
+        timed_predicate_op op(wake_time, wake_predicate, wake_predicate_handle);
+        do_sleep_operation(op);
+    }
+
+    return os_timer->update_and_get_tick();
+}
+
+
+#if MBED_CONF_RTOS_PRESENT
+/* The 32-bit limit is part of the API - we will always wake within 2^32 ticks */
+/* This version is tuned for RTOS use, where the RTOS needs to know the time spent sleeping */
+uint32_t do_timed_sleep_relative(uint32_t wake_delay, bool (*wake_predicate)(void *), void *wake_predicate_handle)
+{
+    uint64_t sleep_start = init_os_timer()->get_tick();
+    // When running with RTOS, the requested delay will be based on the kernel's tick count.
+    // If it missed a tick as entering idle, we should reflect that by moving the
+    // start time back to reflect its current idea of time.
+    // Example: OS tick count = 100, our tick count = 101, requested delay = 50
+    // We need to schedule wake for tick 150, report 50 ticks back to our caller, and
+    // clear the unacknowledged tick count.
+    sleep_start -= os_timer->unacknowledged_ticks();
+
+    uint64_t sleep_finish = do_timed_sleep_absolute(sleep_start + wake_delay, wake_predicate, wake_predicate_handle);
+
+    return (uint32_t) (sleep_finish - sleep_start);
+}
+
+#else
+
+void do_untimed_sleep(bool (*wake_predicate)(void *), void *wake_predicate_handle)
+{
+    untimed_op op(wake_predicate, wake_predicate_handle);
+
+    do_sleep_operation(op);
+}
+
+/* (uint32_t)-1 delay is treated as "wait forever" */
+/* This version is tuned for non-RTOS use, where we don't need to return sleep time, and waiting forever is possible */
+void do_timed_sleep_relative_or_forever(uint32_t wake_delay, bool (*wake_predicate)(void *), void *wake_predicate_handle)
```

Comment:
Warning: this is very much a bike-shedding topic, so I'll endeavor to keep this as short as possible.

To me, the `do_timed_sleep_relative_or_forever` function name is pretty confusing, specifically the `_or_forever` suffix. Is it a relative sleep or a forever sleep? After thinking about it for a bit, I generally assume that I can pass an argument in that makes it a forever sleep (which turns out to be the case), however the name of the function causes enough cognitive dissonance for me to write this comment 😆. Is this a pretty common function name pattern in other libraries?

Other question is how come there isn't an equivalent function for the RTOS implementation? Perhaps having a thread sleep forever doesn't make sense, but I might be missing something else (might be good to add to the commit message?).

My knee-jerk suggested alternative would be have a separate `do_sleep_forever` function OR move the "forever sleep" functionality into either the `do_timed_sleep_relative` or `do_timed_sleep_absolute` function by supplying a constant value (as you have here).

If there's an obvious reason for the current implementation, then please forgive my slowness 😄 But also probably wouldn't hurt to add this reason to the commit message!

Reply:
This is marked as internal, it's not intended for public use, so I'm not too concerned about naming.

Its API is basically "what we need to implement the RTOS API calls in non-RTOS environment smoothly", rather than being any sort of elegant API concept.

I'm trying to be utterly explicit in the name that it accepts the magic wake_delay of `osWaitForever` (`(uint32_t)-1') . That isn't the case for some of the calls it's alongside. And waiting forever is a clearly distinct case.

Reply:
AH totally missed that this is internal, my bad! In that case, be as verbose as you like :)